### PR TITLE
Add suspend and resume API

### PR DESCRIPTION
**Pull request by oleavr**
This allows a native thread to suspend its Duktape thread while for
example calling out to an external blocking API, allowing other native
threads to cooperatively run code on the same heap in the meantime.

It is left up to the application to synchronize its native threads so
only a single native thread executes code at a time, each taking care to
use distinct Duktape threads while doing so.

Related to #834.

### CODE REVIEW

**Comment by svaarala** [comment]
Good stuff, thanks :) Couple of comments and issues:

Use `DUK_MEMMOVE()` instead of `memmove()`; it's abstracted because it needs to be retargeted on some platforms. This causes the code policy check to fail now.

The external state is a 128-byte char array and not an exact match to the internal state. I can't think of a cleaner way right now unfortunately. I have a pending branch for 2.0 where the headers are rearranged so that all internal structures are visible also when the application does `#include "duktape.h"`. This will allow some optimizations (e.g. `duk_get_top()` can be a macro which peeks directly into the internal structure instead of making a call, value pushing can operate directly on the value stack for performance optimized builds) but will also allow that suspend state to be defined without this internal/external inconvenience. We can fix this later.

Please add yourself to AUTHORS.rst; you're already on the second list but not the code/documentation contributor list.

I can add API documentation entries etc as separate pulls. This probably need a small guide note on the threading section and a Wiki Howto for threading which would explain both the default limitations and what suspend/resume will allow (and what it won't).

**Comment by svaarala** [comment]
So as a basic idiom, assuming a mutex for thread exclusion, one would do something like:

```c
duk_ret_t blocking_http_request(duk_context *ctx) {
    duk_thread_state suspend_state;

    /* ... figure out parameters etc ... */

    duk_suspend(ctx, &suspend_state);
    release_mutex();

    /* At this point all Duktape API calls are unsafe, but it's
     * still safe to reference string data pointers. Also buffer
     * data pointers can be OK -if- you know, for your own code,
     * that the buffers are not resized/reconfigured by other
     * threads.
     */

    do_blocking_http_request(uri, ...);

    acquire_mutex();
    duk_resume(ctx, &suspend_state);

    /* ... figure out return values ... */

    return 1;
}
```

**Comment by oleavr** [comment]
Good stuff, thanks :) Couple of comments and issues:

> Use DUK_MEMMOVE() instead of memmove(); it's abstracted because it needs to be retargeted on some platforms. This causes the code policy check to fail now.

Oops! Fixed. I noticed that all existing call-sites perform explicit casts to `void *` and `const void *`, so I did that as well. I'm assuming this is for compilers that try to be smart and end up making bad assumptions about alignment?

> The external state is a 128-byte char array and not an exact match to the internal state. I can't think of a cleaner way right now unfortunately. I have a pending branch for 2.0 where the headers are rearranged so that all internal structures are visible also when the application does #include "duktape.h". This will allow some optimizations (e.g. duk_get_top() can be a macro which peeks directly into the internal structure instead of making a call, value pushing can operate directly on the value stack for performance optimized builds) but will also allow that suspend state to be defined without this internal/external inconvenience. We can fix this later.

Sounds good. I'm also not super-happy about this part, so it'll be great to get rid of this once the internal structures are externally visible.

> Please add yourself to AUTHORS.rst; you're already on the second list but not the code/documentation contributor list.

Done.

> I can add API documentation entries etc as separate pulls. This probably need a small guide note on the threading section and a Wiki Howto for threading which would explain both the default limitations and what suspend/resume will allow (and what it won't).

That would be great.

> Similarly an API testcase is needed, I can try to figure one out.

Awesome!

**Comment by oleavr** [comment]
> So as a basic idiom, assuming a mutex for thread exclusion, one would do something like:
> (...snip...)

+1. Looking great!

**Comment by svaarala** [comment]
The venerable #391 now tracks the changes for 2.0.

**Comment by svaarala** [comment]
Since a new API is introduced any alternatives or variants to the public API syntax should also be considered. The internals don't matter that much.

So, for completeness: another alternative to the current approach where the caller provides a (usually stack allocated) storage would be for `duk_suspend()` to allocate a buffer for the suspend state and push it on the value stack for resume. The call site would then simply be:

    duk_suspend(ctx);

    ....

    duk_resume(ctx);

This would avoid adding a suspend state type and the call site would be simpler. Also if no suspend state is needed later it'd be cleaner to remove it. The downside would be that a suspend/resume would involve a temporary fixed buffer. It's not a big deal but using the stack for the suspend state is more efficient.

Another alternative would be to provide a define for the suspend state size and the caller would then allocate it from the stack like:

    char state[DUK_SUSPEND_STATE_SIZE];

This is worse from a typing point of view so I don't see upsides for this.

Finally, one alternative would be that there was no need for any suspend/resume API calls. That's not achievable on the short term and maybe not even the medium term (one issue being uncaught error handling) so for me this doesn't seem like a relevant option for 2.0.

**Comment by svaarala** [comment]
@andoma If you have time, it'd be great to hear if you could test this out or see if it would work for you.

There's probably going to be a 1.6.0 release from `v1-maintenance` branch and this should be relatively simple to port to 1.x if that makes it easier for you to test.

**Comment by andoma** [comment]
Looks great!

Just want to verify that I do the right thing here.

To do proper suspend/resume the calls need to happen on different duk threads too, right?
So basically when my code jumps into duktape it should look something like this:

```
udata *u = get_udata(superctx);
pthread_mutex_lock(&u->mutex);
duk_context *ctx = duk_get_context(superctx, duk_push_thread(superctx));
// push all the args on ctx
int rc = duk_pcall(ctx, ...);
if(rc) {
...
} 
duk_set_top(superctx, 0); // .. and the created thread is GC'd eventually
pthread_mutex_unlock(&u->mutex);
```

and then inside my blocking I/O call

```
duk_thread_state suspend_state;

udata *u = get_udata(ctx);
const char *url = duk_get_string(ctx);
duk_suspend(ctx, &suspend_state);
pthread_mutex_unlock(&u->mutex);

buffer *b = http_request(url, ...);
pthread_mutex_lock(&u->mutex);
duk_resume(ctx, &suspend_state);

// push return arg

buffer_free(b);
return 1;
```
I assume udata is shared between the thread-contexts using the same heap.

**Comment by andoma** [comment]
Hm, clearly if two of these things execute `duk_set_top(superctx, 0);` will make both threads unreachable by GC so I suppose I need to store the thread-value in some object until duk_pcall() returns

**Comment by svaarala** [comment]
This will need to be explained in the Wiki in some detail because it's not obvious what the caller should do. Couple of observations:

- The reachability of all threads must be guaranteed by the caller at all times.
- The C callstack of the suspended thread must not be unwound while suspended, otherwise one can't resume (and all bets would be off anyway).
- One obvious way to use this feature is when a native thread blocks and another native thread resumes. This implies some form of locking to control access to Duktape API. If the native threads are cooperating rather than pre-emptive, a simpler solution is possible.
- Another less obvious way is for the suspended thread to resume a different Duktape thread (using the same native thread) while the operation is pending. This runs the risk of unbounded C recursion if that thread also suspends, resumes another thread etc; this is ultimately up to the application to control.

**Comment by andoma** [comment]
> @andoma If you have time, it'd be great to hear if you could test this out or see if it would work for you.
> 
> There's probably going to be a 1.6.0 release from v1-maintenance branch and this should be relatively simple to port to 1.x if that makes it easier for you to test.

Tested it (just applied the patch on 1.5.0) and it seems to work perfectly fine for my needs.

**Comment by saghul** [comment]
@oleavr Thanks for making this! I'm the one who opened #834 some time ago. If you don't mind, I have a quick question: if my understanding is correct, with this we would have a model similar to that of CPython or the MRI, a GIL: only one thread can be executing JS code at a time, and syscalls would be some of the "yielding points", where one thread calls `duk_suspend` and waits until the syscall is done, to then call `duk_resume`. How does this model work when multiple threads execute JS code concurrently? CPtyhon, for instance, lets the GIL go every X bytecode instructions, so syscalls are not the only yielding point. How would we go about that?

Thanks!

**Comment by svaarala** [comment]
@saghul There's is no locking here as far as Duktape is concerned - it's up to the application to control access to Duktape API. So there's no overhead or concurrency for interpretation. There are upsides and downsides to this, of course.

**Comment by svaarala** [comment]
@andoma Good to hear :) So now we need to decide which API variant to adopt, and I'll then apply it as a patch to 1.6.0 release for those who don't want to jump into 2.0 just to get this feature.

**Comment by saghul** [comment]
@svaarala I know, let me rephrase :-) In order to allow multiple threads to execute JS code on the same Duktape context one can use these APIs, for example around blocking C code, so another thread can continue with its business. But how would this work if there is no blocking C code around for one thread to suspend execution? Is there a way to register a callback which will be called every so often and allows for suspending execution of JS code wherever it might be? Thanks!

**Comment by svaarala** [comment]
@andoma Good to hear :) So I guess the main open issue is which API variant to adopt, and of course documentation and such.

Regarding the API approaches, one thing that came to mind is that it's important that the C call stack of the suspending thread is not unwound, i.e. a `duk_resume()` must always happen from the same C call stack context as the corresponding `duk_suspend()`.

The explicit state pointer kind makes it harder to do that wrong because the most natural place for the state is a local variable, so it's also quite natural for both `duk_suspend()` and `duk_resume()` to be placed in the same function.

**Comment by svaarala** [comment]
I'll merge this pull (not the alternate approach) pretty soon unless there are some voices to the contrary :) This will then be in master, at least as an experimental feature. API changes before 2.0 release would still be possible, if there are good reasons to do so.

**Comment by oleavr** [comment]
@svaarala Awesome! Thanks a lot for looking into this. Totally agree that this stack-based approach makes it more intuitive to use the API correctly.

**Comment by andoma** [comment]
> @svaarala Awesome! Thanks a lot for looking into this. Totally agree that this stack-based approach makes it more intuitive to use the API correctly.

+1

**Comment by fatcerberus** [comment]
> stack-based approach

*Is* it stack-based?  `duk_suspend()` and `duk_resume()` take a pointer to a state structure.  It never touches the value stack from what I can tell.

**Comment by oleavr** [comment]
> Is it stack-based? duk_suspend() and duk_resume() take a pointer to a state structure. It never touches the value stack from what I can tell.

I meant the *native* stack, sorry, confusing. :-) And yeah, it does touch the value stack because of `value1` / `value2`, but that's the case for both approaches.

**Comment by fatcerberus** [comment]
Ah, that makes more sense then, thanks.

**Comment by kattechou** [comment]
> @svaarala I know, let me rephrase :-) In order to allow multiple threads to execute JS code on the same Duktape context one can use these APIs, for example around blocking C code, so another thread can continue with its business. But how would this work if there is no blocking C code around for one thread to suspend execution? Is there a way to register a callback which will be called every so often and allows for suspending execution of JS code wherever it might be? Thanks!

@saghul Do you get an answer for that? Can you share it? Thanks. (To be clear, liblua has API suited for my need - lua_yield() and lua_resume())

**Comment by saghul** [comment]
I don't rtemember, TBH... I'm no longer working on the project which prompted that question.
